tweet_recommendations/embeddings/sif_embedding.py

```diff
@@ -0,0 +1,130 @@
+                           word_counts.items() if count >= min_word_occurence}
+
+        total_words_count = len(all_words_in_dataset)
+        self.words_weights = {
+            word: smoothing / (smoothing + (count / total_words_count))
+            for word, count
+            in word_counts.items()}
+
+        emb = self._get_weighted_average(x["lemmas"].tolist())
+
+        self._pc = self._compute_pc(emb)
+
+        return self
+
+    def transform(self, x: Union[List[str], str]) -> np.ndarray:
+        """
+        :param x: List of lemmas in a tweet or raw tweet text.
+        :return: Sentence embedding
+        """
+        if isinstance(x, str):
+            x = get_wcrft2_results_for_text(x)
+        x = list(x)
+        sentence_embedding = self._get_sif_embedding(x)
+
+        return sentence_embedding
+
+    def fit_transform(self, x: pd.DataFrame, y=None, **fit_params) -> np.ndarray:
+        self.fit(x, y, **fit_params)
+        embeddings = self._get_sif_embedding(x["lemmas"].tolist())
+        return embeddings
+
+    def _get_weighted_average(self, sentences: List[List[str]]) -> np.ndarray:
+        """
+        Compute the weighted average vectors for given tweet content lemmas.
+        :param sentences: List of tweet content lemmas list.
+        :return: Weighted average of tweet embeddings.
+        """
+        n_samples = len(sentences)
+        emb = np.zeros((n_samples, self.w2v_keyed_vector.vector_size))
+        for i in range(n_samples):
+            if sentences[i]:
+                words_embeddings = []
+                words_weights = []
+                for word in sentences[i]:
+                    words_embeddings.append(self.get_word_vector(word.lower()))
+                    words_weights.append(self.words_weights.get(word.lower(), 0))
+                sentence_word_embeddings = np.array(words_embeddings)
+                sentence_word_weights = np.array(words_weights)
+
+                emb[i, :] = sentence_word_weights.dot(sentence_word_embeddings) / len(
+                    sentence_word_weights)
+        return emb
+
+    def _compute_pc(self, tweets_embeddings: np.ndarray) -> np.ndarray:
+        """
+        Compute the first principal components for given tweet embedding matrix.
+        :param tweets_embeddings: array of tweet embeddings.
+        :return: svd.component_: array of first pca principal component
+        for each tweet embedding.
+        """
+        svd = TruncatedSVD(n_components=1, n_iter=7, random_state=0)
+        svd.fit(tweets_embeddings)
+        return svd.components_
+
+    def _remove_pc(self, tweets_embeddings: np.ndarray) -> np.ndarray:
+        """
+        Remove the projection on the principal components for each tweet embedding.
+        :param tweets_embeddings: array of tweet embeddings.
+        :return: tweets embeddings after removing its projection.
+        """
+        tweet_embeddings_with_first_pc_removed = tweets_embeddings - tweets_embeddings.dot(
```

Comment:
Nie jestem pewien, czy takie usunięcie pierwszej składowej jest poprawne (i z czym to się je). Jeśli masz do tego referencję, skąd to wziąłeś, to byłbym wdzięczny, jakbyś podesłał albo wpisał w komentarzu do metody.

Reply:
Widzę teraz, że w publikacji jest wzorek:
![image](https://user-images.githubusercontent.com/35629146/55656368-f2e25000-57f6-11e9-96c9-8a66aadaf2c6.png)
(pomijam, że nie wiem czy aby na pewno robi to co powinien, ale niech już tak zostanie). Rzecz w tym że nie jestem przekonany, czy to co widzę w kodzie to to samo. W kodzie widzę raczej:
![image](https://user-images.githubusercontent.com/35629146/55656555-9df30980-57f7-11e9-8526-df2e6779e1d2.png)

Reply:
Hm, nie wypowiem się co do samego usuwania składowej, ale jeśli chodzi o wzór, to rzeczywiście coś tu nie gra. Nie sprawdzałem tego dokładnie, ale obecna implementacja wzorowana jest na tej tutaj:
https://github.com/PrincetonML/SIF/blob/master/src/SIF_embedding.py#L39

Reply:
Generalnie to mnie to trochę zastanawia, bo o ile dobrze rozumiem, to jest implementacja dołączona do oryginalnej publikacji i Tak implementacja różni się trochę od tego pseudokodu zamieszczonego w publikacji ... @kacper1095 @mkosturek @dzkb opinie plz

Reply:
No dobra, jak tak jest w oryginalnej implementacji, to nie tak będzie, ale to nie jestem przekonany do niej. Tak czytając sobie, to w sumie implementacja ma dla mnie więcej sensu niż ten wzór, który napisali w paperze.

Reply:
Imho to w publikacji jest tylko takim lewym schematem przetwarzania. Super najsu, cieszę się